packages/grid/x-data-grid/src/components/base/GridBody.tsx

```diff
@@ -57,14 +58,28 @@ function GridBody(props: GridBodyProps) {
 
   const columnHeadersRef = React.useRef<HTMLDivElement>(null);
   const columnsContainerRef = React.useRef<HTMLDivElement>(null);
-  const windowRef = React.useRef<HTMLDivElement>(null);
   const renderingZoneRef = React.useRef<HTMLDivElement>(null);
 
   apiRef.current.columnHeadersContainerElementRef = columnsContainerRef;
   apiRef.current.columnHeadersElementRef = columnHeadersRef;
-  apiRef.current.windowRef = windowRef; // TODO rename, it's not attached to the window anymore
   apiRef.current.renderingZoneRef = renderingZoneRef; // TODO remove, nobody should have access to internal parts of the virtualization
 
+  const targetRef = React.useCallback(
+    (node: HTMLDivElement) => {
+      if (node && apiRef.current.windowRef?.current == null) {
+        apiRef.current.windowRef = { current: node };
+        const triggerGridReadyEvent = debounce(() => apiRef.current.publishEvent('gridReady'), 100);
```

Comment:
> Why not try https://github.com/mui/mui-x/pull/6568#pullrequestreview-1158462141 that doesn't require setTimeout? We should try to avoid setTimeout

I have tried it, but 'gridReady' is getting triggered multiple times. I could not get it to be triggered only 1 time.
```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx b/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
index 782d3db54..b5e7ccf87 100644
--- a/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
+++ b/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
@@ -1,6 +1,7 @@
 import * as React from 'react';
 import * as ReactDOM from 'react-dom';
 import { unstable_useForkRef as useForkRef } from '@mui/utils';
+import { debounce } from '@mui/material/utils';
 import { useGridApiContext } from '../../utils/useGridApiContext';
 import { useGridRootProps } from '../../utils/useGridRootProps';
 import { useGridSelector } from '../../utils/useGridSelector';
@@ -97,6 +98,10 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
     getRowProps,
   } = props;
 
+  const triggerGridReadyEvent = React.useMemo(
+    () => debounce(() => apiRef.current.publishEvent('gridReady'), 300),
+    [apiRef],
+  );
   const columnPositions = useGridSelector(apiRef, gridColumnPositionsSelector);
   const columnsTotalWidth = useGridSelector(apiRef, gridColumnsTotalWidthSelector);
   const cellFocus = useGridSelector(apiRef, gridFocusCellSelector);
@@ -205,7 +210,7 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => { 
     containerWidth,
   ]);
 
-  React.useEffect(() => {
+  React.useLayoutEffect(() => {
     if (disableVirtualization) {
       renderZoneRef.current!.style.transform = `translate3d(0px, 0px, 0px)`;
     } else {
@@ -213,7 +218,13 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
       rootRef.current!.scrollLeft = 0;
       rootRef.current!.scrollTop = 0;
     }
-  }, [disableVirtualization]);
+
+    apiRef.current.subscribeEvent('stateChange', (newState) => {
+      if (newState.rowsMeta.currentPageTotalHeight > 0) {
+        triggerGridReadyEvent();
+      }
+    });
+  }, [disableVirtualization, triggerGridReadyEvent, apiRef]);
```
this is how I tried to implement it.

Reply:
hey @m4theushw 

i was looking at  

https://github.com/mui/mui-x/blob/4d77507ba7aab64c70ab231cb14e386ab96a8ef4/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx#L212-L214

and I think they are getting called too early before the rows are populated, I do not think they have any effect. I have tried `rootRef.current!.scrollTop = 100`.  and it does not do anything

Reply:
hey @m4theushw, I have managed `useEffect` to be called only once. and I have put the code in `useGridVirtualScroller`.